src/Algebra/Solver/IdempotentCommutativeMonoid/Normal.agda

```diff
@@ -0,0 +1,160 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Solver for equations in idempotent commutative monoids
+--
+-- Adapted from Algebra.Solver.CommutativeMonoid
+------------------------------------------------------------------------
+
+{-# OPTIONS --cubical-compatible --safe #-}
+
+open import Algebra.Bundles using (IdempotentCommutativeMonoid)
+
+module Algebra.Solver.IdempotentCommutativeMonoid.Normal
+  {c ℓ} (M : IdempotentCommutativeMonoid c ℓ) where
+
+import Algebra.Properties.CommutativeSemigroup as CSProperties
+open import Data.Bool as Bool using (Bool; true; false; if_then_else_; _∨_)
+open import Data.Fin.Base using (Fin; zero; suc)
+open import Data.Nat.Base using (ℕ)
+open import Data.Vec.Base using (Vec; []; _∷_; lookup; replicate)
+import Data.Vec.Relation.Binary.Pointwise.Inductive as Pointwise
+open import Relation.Binary.Definitions using (DecidableEquality)
+import Relation.Binary.Reasoning.Setoid as ≈-Reasoning
+import Relation.Nullary.Decidable       as Dec
+
+open IdempotentCommutativeMonoid M
+open CSProperties commutativeSemigroup using (x∙yz≈y∙xz)
+open ≈-Reasoning setoid
+
+private
+  variable
+    n : ℕ
+
+
+------------------------------------------------------------------------
+-- Monoid expressions
+
+open import Algebra.Solver.Monoid.Expression rawMonoid public
+  hiding (NormalAPI)
+
+------------------------------------------------------------------------
+-- Normal forms
+
+-- A normal form is a vector of bits (a set).
+
+Normal : ℕ → Set
+Normal n = Vec Bool n
+
+-- The semantics of a normal form.
+
+⟦_⟧⇓ : Normal n → Env n → Carrier
+⟦ []    ⟧⇓ _ = ε
+⟦ b ∷ v ⟧⇓ (a ∷ ρ) = if b then a ∙ (⟦ v ⟧⇓ ρ) else (⟦ v ⟧⇓ ρ)
+
+-- We can decide if two normal forms are /syntactically/ equal.
+
+infix 5 _≟_
+
+_≟_ : DecidableEquality (Normal n)
+nf₁ ≟ nf₂ = Dec.map Pointwise-≡↔≡ (decidable Bool._≟_ nf₁ nf₂)
+  where open Pointwise using (Pointwise-≡↔≡; decidable)
+
+------------------------------------------------------------------------
+-- Constructions on normal forms
+
+-- The empty set.
+
+empty : Normal n
+empty = replicate _ false
+
+-- A singleton set.
+
+sg : (i : Fin n) → Normal n
+sg zero    = true ∷ empty
+sg (suc i) = false ∷ sg i
+
+-- The composition of normal forms.
+infixr 10 _•_
+
+_•_ : (v w : Normal n) → Normal n
+[]      • []      = []
```

Comment:
`zipWith` ?

Reply:
!!! nice observation !!!

Reply:
(I'd not attempted to actually analyse existing code, only move it around...)